src/hotspot/cpu/aarch64/register_aarch64.hpp

```diff
@@ -232,36 +238,41 @@ CONSTANT_REGISTER_DECLARATION(FloatRegister, z31    , (31));
 
 
 class PRegisterImpl;
-typedef PRegisterImpl* PRegister;
-inline PRegister as_PRegister(int encoding) {
-  return (PRegister)(intptr_t)encoding;
-}
+typedef const PRegisterImpl* PRegister;
+const PRegister as_PRegister(int encoding);
 
 // The implementation of predicate registers for the architecture
 class PRegisterImpl: public AbstractRegisterImpl {
+  friend const PRegister as_PRegister(int encoding);
+
+private:
+  static constexpr PRegister first();
+  static constexpr PRegister invalid() { return first() + number_of_declared_registers; }
+
  public:
   enum {
     number_of_registers = 16,
+    number_of_declared_registers = 16,
```

Comment:
Is there a semantic difference between number of registers and declared registers for float and P registers? Otherwise, having two constants for the same thing makes the code less clear and maybe more error prone.

Reply:
Perhaps, but I'm trying to do something really general. There is _sometimes_ a difference between the number of physical registers and the number of named registers, and that's what I'm trying to capture here.